File: src/dep_health_scanner/lockfile.py
from __future__ import annotations

import json
import re
from pathlib import Path
from typing import List, Optional

from .models import Dependency, Ecosystem, Lockfile


class LockfileDetector:
    PATTERNS: List[tuple[str, Ecosystem]] = [
        ("package-lock.json", Ecosystem.NPM),
        ("yarn.lock", Ecosystem.NPM),
        ("pnpm-lock.yaml", Ecosystem.NPM),
        ("Cargo.lock", Ecosystem.CARGO),
        ("Pipfile.lock", Ecosystem.PIP),
        ("poetry.lock", Ecosystem.PIP),
        ("go.mod", Ecosystem.GO),
    ]

    @classmethod
    def detect(cls, root: Path) -> Optional[Lockfile]:
        for filename, eco in cls.PATTERNS:
            path = root / filename
            if path.exists():
                return cls._parse(path, eco)
        return None

    @classmethod
    def _parse(cls, path: Path, ecosystem: Ecosystem) -> Lockfile:
        if ecosystem == Ecosystem.NPM:
            deps = cls._parse_npm(path)
        elif ecosystem == Ecosystem.CARGO:
            deps = cls._parse_cargo(path)
        elif ecosystem == Ecosystem.PIP:
            deps = cls._parse_pip(path)
        elif ecosystem == Ecosystem.GO:
            deps = cls._parse_go_mod(path)
        else:
            raise ValueError(f"Unsupported ecosystem: {ecosystem}")
        return Lockfile(path=path, ecosystem=ecosystem, dependencies=deps)

    @classmethod
    def _parse_npm(cls, path: Path) -> List[Dependency]:
        data = json.loads(path.read_text())
        deps: List[Dependency] = []
        # package-lock.json v2+
        packages = data.get("packages") or data.get("dependencies") or {}
        if isinstance(packages, dict):
            for key, info in packages.items():
                if not key or not isinstance(info, dict):
                    continue
                version = info.get("version")
                if not version:
                    continue
                name = (
                    key.split("/")[0]
                    if not key.startswith("node_modules/")
                    else key.replace("node_modules/", "")
                )
                deps.append(
                    Dependency(
                        name=name,
                        version=str(version),
                        ecosystem=Ecosystem.NPM,
                        transitive="node_modules/" in key,
                    )
                )
        return deps

    @classmethod
    def _parse_cargo(cls, path: Path) -> List[Dependency]:
        data = json.loads(path.read_text())
        packages = data.get("package") or data.get("packages") or []
        deps: List[Dependency] = []
        if isinstance(packages, list):
            for pkg in packages:
                deps.append(
                    Dependency(
                        name=pkg["name"],
                        version=pkg["version"],
                        ecosystem=Ecosystem.CARGO,
                        source=pkg.get("source"),
                        transitive=True,
                    )
                )
        return deps

    @classmethod
    def _parse_pip(cls, path: Path) -> List[Dependency]:
        # Pipfile.lock is JSON; poetry.lock is TOML. Handle both.
        if path.suffix == ".toml":
            import tomllib

            with path.open("rb") as f:
                data = tomllib.load(f)
            packages = data.get("package", [])
            deps = [
                Dependency(name=p["name"], version=p["version"], ecosystem=Ecosystem.PIP)
                for p in packages
            ]
            return deps
        data = json.loads(path.read_text())
        deps: List[Dependency] = []
        for section in ("default", "develop"):
            if section in data and isinstance(data[section], dict):
                for name, info in data[section].items():
                    version = info.get("version")
                    if version:
                        deps.append(
                            Dependency(
                                name=name,
                                version=version,
                                ecosystem=Ecosystem.PIP,
                                transitive=False,
                            )
                        )
        return deps

    @classmethod
    def _parse_go_mod(cls, path: Path) -> List[Dependency]:
        lines = path.read_text().splitlines()
        deps: List[Dependency] = []
        inside_block = False

        for raw in lines:
            stripped = raw.strip()

            if inside_block:
                if stripped.startswith(")"):
                    inside_block = False
                    continue
                if not stripped or stripped.startswith("//"):
                    continue
                name, version, *_ = re.split(r"\s+", stripped)
                is_indirect = "// indirect" in stripped
                deps.append(
                    Dependency(
                        name=name,
                        version=version.lstrip("v"),
                        ecosystem=Ecosystem.GO,
                        transitive=is_indirect,
                    )
                )
                continue

            if not stripped.startswith("require"):
                continue

            block = stripped[len("require") :].strip()
            if block.startswith("("):
                inside_block = True
                # drop the opening paren on the same line
                after_paren = block[1:].strip()
                if after_paren and not after_paren.startswith("//"):
                    name, version, *_ = re.split(r"\s+", after_paren)
                    is_indirect = "// indirect" in after_paren
                    deps.append(
                        Dependency(
                            name=name,
                            version=version.lstrip("v"),
                            ecosystem=Ecosystem.GO,
                            transitive=is_indirect,
                        )
                    )
                continue

            name, version, *_ = re.split(r"\s+", block)
            is_indirect = "// indirect" in block
            deps.append(
                Dependency(
                    name=name,
                    version=version.lstrip("v"),
                    ecosystem=Ecosystem.GO,
                    transitive=is_indirect,
                )
            )

        return deps

File: src/lockfile.rs
use super::dependency::{Dependency, Ecosystem, Lockfile};
use super::Error;
use glob::glob;
use serde::Deserialize;
use std::fs;
use std::path::{Path, PathBuf};

pub struct LockfileDetector;

impl LockfileDetector {
    pub fn detect(root: &Path) -> Result<Option<Lockfile>> {
        let patterns = vec![
            ("package-lock.json", Ecosystem::Npm),
            ("yarn.lock", Ecosystem::Npm),
            ("pnpm-lock.yaml", Ecosystem::Npm),
            ("Cargo.lock", Ecosystem::Cargo),
            ("Pipfile.lock", Ecosystem::Pip),
            ("poetry.lock", Ecosystem::Pip),
            ("go.sum", Ecosystem::Go),
        ];

        for (filename, eco) in patterns {
            let path = root.join(filename);
            if path.exists() {
                let lockfile = Self::parse(&path, eco)?;
                return Ok(Some(lockfile));
            }
        }

        // Recursive search for lockfiles (limited depth)
        for entry in glob(root.join("**/package-lock.json").to_str().unwrap()).unwrap() {
            if let Ok(path) = entry {
                let lockfile = Self::parse(&path, Ecosystem::Npm)?;
                return Ok(Some(lockfile));
            }
        }

        Ok(None)
    }

    pub fn detect_all(root: &Path) -> Result<Vec<Lockfile>> {
        let mut lockfiles = Vec::new();
        let patterns = vec![
            ("package-lock.json", Ecosystem::Npm),
            ("Cargo.lock", Ecosystem::Cargo),
            ("Pipfile.lock", Ecosystem::Pip),
        ];

        for (filename, eco) in patterns {
            for entry in glob(root.join(format!("**/{}", filename)).to_str().unwrap()).unwrap() {
                if let Ok(path) = entry {
                    if let Ok(lf) = Self::parse(&path, eco) {
                        lockfiles.push(lf);
                    }
                }
            }
        }

        if lockfiles.is_empty() {
            return Err(Error::NoLockfileFound);
        }

        Ok(lockfiles)
    }

    fn parse(path: &Path, ecosystem: Ecosystem) -> Result<Lockfile> {
        let content = fs::read_to_string(path)?;
        let deps = match ecosystem {
            Ecosystem::Npm => Self::parse_npm(&content)?,
            Ecosystem::Cargo => Self::parse_cargo(&content)?,
            Ecosystem::Pip => Self::parse_pip(&content)?,
            _ => return Err(Error::UnsupportedFormat(format!("{:?}", ecosystem))),
        };

        Ok(Lockfile {
            path: path.to_path_buf(),
            ecosystem,
            dependencies: deps,
        })
    }

    fn parse_npm(content: &str) -> Result<Vec<Dependency>> {
        #[derive(Deserialize)]
        struct PackageLock {
            packages: Option<serde_json::Value>,
            dependencies: Option<serde_json::Value>,
        }

        let lock: PackageLock = serde_json::from_str(content)?;
        let mut deps = Vec::new();

        if let Some(packages) = lock.packages {
            if let Some(obj) = packages.as_object() {
                for (key, value) in obj {
                    if key == "" {
                        continue;
                    }
                    if let Some(version) = value.get("version").and_then(|v| v.as_str()) {
                        let name = key.split('/').next().unwrap_or(key).to_string();
                        let transitive = key.contains("node_modules/");
                        deps.push(Dependency {
                            name,
                            version: version.to_string(),
                            ecosystem: Ecosystem::Npm,
                            source: None,
                            transitive,
                            dependencies: Vec::new(),
                        });
                    }
                }
            }
        } else if let Some(deps) = lock.dependencies {
            Self::extract_npm_deps(&deps, &mut deps, false);
        }

        Ok(deps)
    }

    fn extract_npm_deps(value: &serde_json::Value, deps: &mut Vec<Dependency>, transitive: bool) {
        if let Some(obj) = value.as_object() {
            for (name, info) in obj {
                if let Some(version) = info.get("version").and_then(|v| v.as_str()) {
                    deps.push(Dependency {
                        name: name.clone(),
                        version: version.to_string(),
                        ecosystem: Ecosystem::Npm,
                        source: None,
                        transitive,
                        dependencies: Vec::new(),
                    });
                }
                if let Some(sub) = info.get("dependencies") {
                    Self::extract_npm_deps(sub, deps, true);
                }
            }
        }
    }

    fn parse_cargo(content: &str) -> Result<Vec<Dependency>> {
        #[derive(Deserialize)]
        struct CargoLock {
            package: Option<Vec<CargoPackage>>,
        }

        #[derive(Deserialize)]
        struct CargoPackage {
            name: String,
            version: String,
            source: Option<String>,
        }

        let lock: CargoLock = serde_json::from_str(content)?;
        let mut deps = Vec::new();

        if let Some(packages) = lock.package {
            for pkg in packages {
                deps.push(Dependency {
                    name: pkg.name,
                    version: pkg.version,
                    ecosystem: Ecosystem::Cargo,
                    source: pkg.source,
                    transitive: true,
                    dependencies: Vec::new(),
                });
            }
        }

        Ok(deps)
    }

    fn parse_pip(content: &str) -> Result<Vec<Dependency>> {
        #[derive(Deserialize)]
        struct PipfileLock {
            default: Option<std::collections::HashMap<String, serde_json::Value>>,
            develop: Option<std::collections::HashMap<String, serde_json::Value>>,
        }

        let lock: PipfileLock = serde_json::from_str(content)?;
        let mut deps = Vec::new();

        let add_section =
            |section: Option<&std::collections::HashMap<String, serde_json::Value>>,
             deps: &mut Vec<Dependency>| {
                if let Some(map) = section {
                    for (name, info) in map {
                        if let Some(version) = info.get("version").and_then(|v| v.as_str()) {
                            deps.push(Dependency {
                                name: name.clone(),
                                version: version.to_string(),
                                ecosystem: Ecosystem::Pip,
                                source: None,
                                transitive: false,
                                dependencies: Vec::new(),
                            });
                        }
                    }
                }
            };

        add_section(lock.default.as_ref(), &mut deps);
        add_section(lock.develop.as_ref(), &mut deps);

        Ok(deps)
    }
}
